gnc/perception/sub8_pcl/nodes/buoy.cc

```diff
@@ -0,0 +1,291 @@
+                                             sub8_msgs::VisionRequest::Response &resp) {
+  std::string tf_frame;
+  Eigen::Vector3f position;
+
+  if (!got_cloud) {
+    // Failure, yo!
+    return false;
+  }
+  computing = true;
+  tf_frame = cam_model.tfFrame();
+  determine_buoy_position(cam_model, current_image, current_cloud, position);
+
+  tf::pointEigenToMsg(position.cast<double>(), resp.pose.pose.position);
+  resp.pose.header.frame_id = tf_frame;
+  rviz.visualize_buoy(resp.pose.pose, tf_frame);
+
+#ifdef VISUALIZE
+  cv::Point2d cv_pt_uv =
+      cam_model.project3dToPixel(cv::Point3f(position.x(), position.y(), position.z()));
+  cv::circle(current_image, cv_pt_uv, 7, cv::Scalar(0, 20, 240), 2);
+  cv::imshow("input", current_image);
+  cv::waitKey(50);
+#endif
+  computing = false;
+
+  return true;
+}
+
+void Sub8BuoyDetector::compute_loop(const ros::TimerEvent &timer_event) {
+#ifdef VISUALIZE
+  viewer->spinOnce();
+#endif
+}
+
+void Sub8BuoyDetector::determine_buoy_position(
+    const image_geometry::PinholeCameraModel &camera_model, const cv::Mat &image_raw,
+    const sub::PointCloudT::Ptr &point_cloud_raw, Eigen::Vector3f &center) {
+  cv::Mat image_hsv;
+  cv::Mat image_thresh;
+
+  // believe it or not, this is on purpose (So blue replaces red in HSV)
+  cv::cvtColor(image_raw, image_hsv, CV_RGB2HSV);
+
+  cv::Point2d pt_cv_2d(250, 250);
+  sub::PointXYZT pcl_pt_3d;
+  pcl_pt_3d = sub::project_uv_to_cloud(*point_cloud_raw, pt_cv_2d, camera_model);
+
+  // Reprojection (Useful for visualization)
+  cv::Point2d cv_pt_uv =
+      cam_model.project3dToPixel(cv::Point3f(pcl_pt_3d.x, pcl_pt_3d.y, pcl_pt_3d.z));
+
+  // Threshold -- > This is what must be replaced with better 2d vision
+  cv::inRange(image_hsv, cv::Scalar(105, 135, 135), cv::Scalar(120, 255, 255), image_thresh);
+  std::vector<sub::Contour> contours;
+  std::vector<cv::Vec4i> hierarchy;
+  cv::findContours(image_thresh.clone(), contours, hierarchy, CV_RETR_EXTERNAL,
+                   CV_CHAIN_APPROX_SIMPLE, cv::Point(0, 0));
+  // TODO: ^^^ Make into function ^^^
+
+  // Not returning segmented point cloud for now
+  // sub::PointCloudT::Ptr segmented_cloud(new sub::PointCloudT());
+  // Vector of Eigen-vectors
+  std::vector<Eigen::Vector3f, Eigen::aligned_allocator<Eigen::Vector3f> > buoy_centers;
+
+  // Loop through the contours we found and find the positions
+  for (size_t i = 0; i < contours.size(); i++) {
+    cv::Point contour_centroid = sub::contour_centroid(contours[i]);
+    std::vector<cv::Point> approx;
+
+    // Magic num: 5
+    cv::approxPolyDP(contours[i], approx, 5, true);
+    double area = cv::contourArea(approx);
+
+    // Ignore small regions
+    if (area < 100) {
+      continue;
+    }
+    sub::PointXYZT centroid_projected =
+        sub::project_uv_to_cloud(*point_cloud_raw, contour_centroid, camera_model);
+    buoy_centers.push_back(sub::point_to_eigen(centroid_projected));
```

Comment:
Good catch. I can see this causing problems, if we write missions with the assumption that the returned point is the center of the buoy. I've added a few lines to slide the point along the vector by one buoy radius, to get an approximate center